engineio/src/transports_async/mod.rs

```diff
@@ -0,0 +1,94 @@
+use std::{borrow::Cow, str::from_utf8, sync::Arc};
+
+use crate::{error::Result, Error, Packet, PacketId};
+use bytes::{BufMut, Bytes, BytesMut};
+use futures_util::{
+    stream::{SplitSink, SplitStream},
+    SinkExt, StreamExt,
+};
+use tokio::{net::TcpStream, sync::Mutex};
+use tokio_tungstenite::{MaybeTlsStream, WebSocketStream};
+use tungstenite::Message;
+
+pub(crate) mod websocket;
+pub(crate) mod websocket_secure;
+
+/// A general purpose asynchronous websocket transport type. Holds
+/// the sender and receiver stream of a websocket connection
+/// and implements the common methods `update`, `poll` and `emit`.
+pub(crate) struct AsyncWebsocketGeneralTransport {
+    sender: Arc<Mutex<SplitSink<WebSocketStream<MaybeTlsStream<TcpStream>>, Message>>>,
+    receiver: Arc<Mutex<SplitStream<WebSocketStream<MaybeTlsStream<TcpStream>>>>>,
+}
```

Comment:
Seems a little odd to have this in the mod for all async transports, but I suppose it's shared between websocket and websocket_secure. 

websocket and websocket_secure also appear to be mostly boilerplate

Reply:
They are indeed, the General transport abstracts over those two, the convenient type here is `MaybeTlsStream` allowing for a coexistence. I will take a look on how both of them can be joined...